### DISCUSSION TITLE
🐈‍⬛ Github을 통한 작업 흐름을 알아보자

### DISCUSSION THREAD

**Discussion by OneMoreThink** [DeveloperAcademy-POSTECH/.github#29] [Idea]
다들 안녕하신지요? 아임 파인입니다. 
Bridge기간 동안 다가올 MC3를 대비해서 Github을 이용해 팀 협업을 하는 흐름을 한번 정리해보았습니다. 
여기저기 쏘다니며 끌어모아 일반적으로 작성해서 너무 Too much일 수 있습니다. 
향후 각자의 팀 상황에 맞게 유동적으로 적용할 필요가 있어보입니다 😸
내용중 오류나 추가할 내용에 대한 Comment 기다리고 있겠습니다. 다들 즐거운 🌉 되길 

##  작업할 내용에 대해 먼저 issue를 발행한다. 
- New issue 클릭
![스크린샷 2024-06-24 오전 12 04 53](https://github.com/DeveloperAcademy-POSTECH/.github/assets/121593683/59f2fc4e-4cde-4350-a026-f0a9ee89241d)
- issue 내용 작성하기 
  - Add a title : [ 이슈 종류 ] 이슈 이름  -> ex ) `[feat] 로그인 화면 추가`
  - Add a description : issue에 대한 구체적인 설명
  - Assigness : 해당 issue를 담당할 사람
  - Labels : 해당 issue의 성격에 따라 설정 
  - Projects : 연결할 Project 설정
  - MileStone : 연결할 MileStone 설정 
- Submit new issue 
![스크린샷 2024-06-24 오전 12 07 37](https://github.com/DeveloperAcademy-POSTECH/.github/assets/121593683/3aedb2fe-8d5d-4549-af3b-b171d66723cb)
- issue 번호가 발급  -> 해당 issue 번호는 브랜치 및 커밋에 사용 
  - 다음 예제로 생성된 이슈번호는 #1
![image](https://github.com/DeveloperAcademy-POSTECH/.github/assets/121593683/468a1ce8-e148-48e2-9e66-a1e51a5f6299)

## 발행한 issue를 작업할 branch를 만든다. 
- 발급된 issue에서 Create a branch 선택하기 
![스크린샷 2024-06-24 오전 12 17 20](https://github.com/DeveloperAcademy-POSTECH/.github/assets/121593683/f2becb64-8f88-4004-992b-74262d648271)

- Branch name 설정하기 : branch 이름  : `분류/#이슈번호-작업할 내용`
- 작업할 Repository 선택
- branch를 낼 branch 선택하기 (어디로부터 분기를 낼건지?)
- Create branch 
![스크린샷 2024-06-24 오전 12 19 52](https://github.com/DeveloperAcademy-POSTECH/.github/assets/121593683/5dc56d1d-fecc-45f6-8124-87fc122fda27)

> 현재 까지 원격에서 발행한 issue의 branch가 생성된 상태 

## 해당 branch를 local에 들고와서 작업을 시작한다. 
- Intergrate > Refresh File Status 선택 
![image 2](https://github.com/DeveloperAcademy-POSTECH/.github/assets/121593683/52f19e78-aa43-4578-9065-a381d8091c1f)
- Remotes/origin에 방금 원격에서 만든 branch가 나타남 
<img width="1391" alt="스크린샷 2024-06-24 오전 12 20 44" src="https://github.com/DeveloperAcademy-POSTECH/.github/assets/121593683/700a9669-38c4-4391-bdb5-f79ce77f24c8">
- 해당 branch에 Switch 
<img width="1397" alt="스크린샷 2024-06-24 오전 12 21 01" src="https://github.com/DeveloperAcademy-POSTECH/.github/assets/121593683/ea6940a3-489f-4878-97d1-be31d7a72bda">
- Local에도 해당 branch가 생김 -> 이제 해당 branch에 작업 시작 
<img width="316" alt="스크린샷 2024-06-24 오전 12 21 14" src="https://github.com/DeveloperAcademy-POSTECH/.github/assets/121593683/69981d0a-800c-4834-b6d5-fcbc285fa8f8">

## 커밋 남기기 
- 현재 내가 작업하는 branch를 먼저 확인하고  작업을 시작합니다. 
![image 3](https://github.com/DeveloperAcademy-POSTECH/.github/assets/121593683/0cd48e67-2e5a-4125-a8bf-48b1cbdc92f7)

- commit 남기기 
  - commit할 내용 Stage
  - commit message 남기기  : [#이슈번호] 이슈종류  : 변경 내용 
  - commit 
<img width="1399" alt="스크린샷 2024-06-24 오전 12 49 57" src="https://github.com/DeveloperAcademy-POSTECH/.github/assets/121593683/996f62f8-68e7-48fd-803d-9017c97c295c">
<img width="1400" alt="스크린샷 2024-06-24 오전 12 50 10" src="https://github.com/DeveloperAcademy-POSTECH/.github/assets/121593683/20f0f78c-e793-4267-b6c6-98f789f626fa">
- 현재 local branch가 하나 앞섰으므로 원격의 branch에 push 해주기 
<img width="1401" alt="스크린샷 2024-06-24 오전 12 50 19" src="https://github.com/DeveloperAcademy-POSTECH/.github/assets/121593683/9477ce4a-ec8b-40fd-97c0-09dc9952c7c7">

## Pull Reqeust 만들기 
- 하나의 issue에 대한 작업을 하면서 해당 branch에 commit을 쌓아 나가다 
- 해당 issue에 대한 작업이 완료되었을 때 해당 branch를 main (develop 등등 team branch 전략에 따라 달라짐)에 merge하기 위해서 pull reqeust를 만들어야한다. 

- 로컬에서 push를 하고 원격에 들어가보면 커밋이 새로 생긴 branch에 대한 pull request를 만들것이냐고 물어본다. 
![스크린샷 2024-06-24 오전 12 54 32](https://github.com/DeveloperAcademy-POSTECH/.github/assets/121593683/ef613bc5-15af-457c-8233-938db26064bb)

- pull request 내용 작성 및 변경된 내용 확인하기 
![image 4](https://github.com/DeveloperAcademy-POSTECH/.github/assets/121593683/0f4b1fa9-b2be-4d48-bac9-4b119c6f08df)
![image 5](https://github.com/DeveloperAcademy-POSTECH/.github/assets/121593683/1e2addc6-a446-4761-b742-0787268a44bd)

- Create pull request 
![스크린샷 2024-06-24 오전 12 54 32 2](https://github.com/DeveloperAcademy-POSTECH/.github/assets/121593683/33956bd1-19eb-4fe8-a81f-32d8f332a990)

- Team Rule에따라 Code Review를 진행후 해당 branch merge를 Confirm 합니다. 
- Local에 돌아가서 mainBranch에 Switch하고 Pull을 하면 원격에서 merge된 내용이 반영됩니다. 

![image 6](https://github.com/DeveloperAcademy-POSTECH/.github/assets/121593683/50393c9e-c5d5-4edb-b03e-2cccd6368471)
![image 7](https://github.com/DeveloperAcademy-POSTECH/.github/assets/121593683/f5db1b81-1d5c-47c0-b565-d6233eb1274f)
![image 8](https://github.com/DeveloperAcademy-POSTECH/.github/assets/121593683/b0379340-5672-47cd-b164-6c4f77a6ec6c)

#  Github Convention

## Issue 종류
- 현재 발행하려는 issue의 종류가 무엇인지 확인합니다. 
<table>
  <thead>
    <tr>
      <th>이슈 종류</th>
      <th>내용</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>feat</td>
      <td>새로운 기능에 대한 이슈</td>
    </tr>
    <tr>
      <td>fix</td>
      <td>버그 수정에 대한 이슈</td>
    </tr>
    <tr>
      <td>build</td>
      <td>빌드 관련 파일 수정 / 모듈 설치 또는 삭제에 대한 이슈</td>
    </tr>
    <tr>
      <td>chore</td>
      <td>그 외 자잘한 수정에 대한 이슈</td>
    </tr>
    <tr>
      <td>ci</td>
      <td>ci 관련 설정 수정에 대한 이슈</td>
    </tr>
    <tr>
      <td>docs</td>
      <td>문서 수정에 대한 이슈</td>
    </tr>
    <tr>
      <td>style</td>
      <td>코드 스타일 혹은 포맷 등에 관한 이슈</td>
    </tr>
    <tr>
      <td>refactor</td>
      <td>코드 리팩토링에 대한 이슈</td>
    </tr>
    <tr>
      <td>test</td>
      <td>테스트 코드 수정에 대한 이슈</td>
    </tr>
    <tr>
      <td>perf</td>
      <td>성능 개선에 대한 이슈</td>
    </tr>
  </tbody>
</table>

## Issue Template
- 선택한 issue에 맞는 Template을 선택합니다. 
### feat (새로운 기능에 대한 이슈)

```
## 기능 요청

### 설명
- 새로운 로그인 화면을 추가합니다.

### 필요 이유
- 사용자들이 앱에 로그인할 수 있도록 하기 위함입니다.

### 구현 방법
- SwiftUI를 사용하여 로그인 화면을 구성합니다.
- 이메일과 비밀번호 입력 필드, 로그인 버튼을 추가합니다.
- 로그인 버튼을 누르면 Firebase Auth를 사용하여 인증을 처리합니다.

### 관련 자료
- [SwiftUI Tutorials](https://developer.apple.com/tutorials/swiftui)
- [Firebase Auth Documentation](https://firebase.google.com/docs/auth)

### 추가 정보
- UI 디자인은 Figma에서 확인할 수 있습니다.
- 로그인 화면은 앱의 루트 뷰로 설정해야 합니다.
```

### fix (버그 수정에 대한 이슈)

```
## 버그 보고

### 설명
- 로그인 시 비밀번호 입력 필드의 보안 텍스트가 정상적으로 표시되지 않습니다.

### 재현 방법
1. 앱을 실행합니다.
2. 로그인 화면으로 이동합니다.
3. 비밀번호 입력 필드에 텍스트를 입력합니다.

### 예상 동작
- 비밀번호 입력 필드에 입력한 텍스트가 점으로 표시되어야 합니다.

### 실제 동작
- 비밀번호 입력 필드에 입력한 텍스트가 일반 텍스트로 표시됩니다.

### 스크린샷
- (스크린샷 첨부)

### 환경
- OS: iOS 15
- 기기: iPhone 13
- 버전: 1.0.0

### 추가 정보
- 비밀번호 입력 필드에 `secureField` 수정이 필요합니다.
```

### build (빌드 관련 파일 수정 / 모듈 설치 또는 삭제에 대한 이슈)

```
## 빌드 요청

### 설명
- Firebase SDK를 프로젝트에 추가합니다.

### 필요 이유
- 로그인 기능 구현을 위해 Firebase Auth를 사용해야 합니다.

### 변경 사항
- `Podfile`에 Firebase Auth 추가
- `pod install` 명령어 실행

### 테스트 방법
- Firebase Auth가 정상적으로 설치되었는지 확인합니다.
- 빌드가 성공적으로 완료되는지 확인합니다.

### 추가 정보
- Firebase 프로젝트 설정은 [Firebase Console](https://console.firebase.google.com/)에서 확인할 수 있습니다.
```

### chore (그 외 자잘한 수정에 대한 이슈)

```
## 작업 요청

### 설명
- 프로젝트 디렉토리 구조를 정리합니다.

### 필요 이유
- 코드의 가독성을 높이고 유지보수성을 향상시키기 위함입니다.

### 변경 사항
- View, ViewModel, Model 디렉토리를 추가하고 파일을 이동합니다.
- 불필요한 파일 삭제

### 추가 정보
- 새로운 디렉토리 구조는 [디렉토리 구조 가이드](#) 문서를 참고하세요.
```

### ci (CI 관련 설정 수정에 대한 이슈)

```
## CI 설정 요청

### 설명
- GitHub Actions 설정을 추가하여 CI 파이프라인을 구성합니다.

### 필요 이유
- 코드 변경 시 자동으로 빌드 및 테스트를 실행하여 품질을 유지하기 위함입니다.

### 변경 사항
- `.github/workflows/ci.yml` 파일을 추가하여 CI 파이프라인을 설정합니다.

### 테스트 방법
- PR 생성 시 GitHub Actions에서 빌드 및 테스트가 자동으로 실행되는지 확인합니다.

### 추가 정보
- CI 설정 예제는 [GitHub Actions Documentation](https://docs.github.com/en/actions)에서 확인할 수 있습니다.
```

### docs (문서 수정에 대한 이슈)

```
## 문서 수정 요청

### 설명
- README.md 파일을 업데이트하여 프로젝트 설정 및 빌드 방법을 추가합니다.

### 필요 이유
- 새로운 개발자들이 프로젝트를 쉽게 설정하고 빌드할 수 있도록 하기 위함입니다.

### 변경 사항
- 프로젝트 설정 방법 섹션 추가
- 빌드 및 실행 방법 섹션 추가

### 추가 정보
- Firebase 설정 방법은 [Firebase Auth Documentation](https://firebase.google.com/docs/auth)에서 확인할 수 있습니다.
```

### style (코드 스타일 혹은 포맷 등에 관한 이슈)

```
## 코드 스타일 수정 요청

### 설명
- SwiftLint를 적용하여 코드 스타일을 일관되게 유지합니다.

### 필요 이유
- 코드의 가독성을 높이고 유지보수성을 향상시키기 위함입니다.

### 변경 사항
- `SwiftLint`를 설치하고 설정 파일을 추가합니다.
- 기존 코드에 linting을 적용합니다.

### 추가 정보
- SwiftLint 설정 예제는 [SwiftLint GitHub](https://github.com/realm/SwiftLint)에서 확인할 수 있습니다.
```

### refactor (코드 리팩토링에 대한 이슈)

```
## 리팩토링 요청

### 설명
- 로그인 로직을 ViewModel로 분리합니다.

### 필요 이유
- MVVM 아키텍처를 따르기 위해 로직을 ViewModel로 분리하여 코드의 유지보수성을 향상시키기 위함입니다.

### 변경 사항
- View에서 로그인 로직을 제거하고 ViewModel에 추가합니다.
- View와 ViewModel 간의 바인딩을 설정합니다.

### 추가 정보
- MVVM 아키텍처에 대한 설명은 [MVVM 패턴 문서](#)에서 확인할 수 있습니다.
```

### test (테스트 코드 수정에 대한 이슈)

```
## 테스트 코드 수정 요청

### 설명
- 로그인 기능에 대한 단위 테스트를 추가합니다.

### 필요 이유
- 로그인 기능의 안정성을 보장하기 위함입니다.

### 변경 사항
- ViewModel의 로그인 로직에 대한 단위 테스트 추가
- 성공 및 실패 케이스 테스트

### 추가 정보
- 단위 테스트 작성 방법은 [Swift Testing Documentation](https://developer.apple.com/documentation/xctest)에서 확인할 수 있습니다.
```

### perf (성능 개선에 대한 이슈)

```
## 성능 개선 요청

### 설명
- 리스트 뷰에서 이미지 로딩 성능을 개선합니다.

### 필요 이유
- 리스트 뷰의 스크롤 성능을 향상시키기 위함입니다.

### 변경 사항
- 비동기 이미지 로딩 및 캐싱을 적용합니다.
- `ImageLoader` 유틸리티를 사용하여 이미지 로딩 최적화

### 테스트 방법
- 대용량 데이터로 리스트 뷰를 테스트하여 스크롤 성능을 확인합니다.

### 추가 정보
- 이미지 로딩 최적화 방법은 [SwiftUI Image Loading](https://developer.apple.com/documentation/swiftui/image)에서 확인할 수 있습니다.
```

### feat (새로운 기능에 대한 커밋)

```
[#3] feat: 홈 화면 UI 추가 - 홈 화면에 필요한 UI 컴포넌트 추가
```

### fix (버그 수정에 대한 커밋)

```
[#56] fix: 비밀번호 확인 오류 수정 - 비밀번호 확인 로직 및 에러 메시지 수정
```

### build (빌드 관련 파일 수정 / 모듈 설치 또는 삭제에 대한 커밋)

```
[#34] build: Firebase SDK 추가 - Podfile에 Firebase Auth 추가 및 pod install 실행
```

### chore (그 외 자잘한 수정에 대한 커밋)

```
[#12] chore: 프로젝트 디렉토리 정리 - 불필요한 파일 삭제 및 폴더 구조 정리
```

### ci (CI 관련 설정 수정에 대한 커밋)

```
[#78] ci: GitHub Actions 설정 추가 - 테스트 워크플로우에 새로운 환경 변수 추가
```

### docs (문서 수정에 대한 커밋)

```
[#45] docs: README.md 업데이트 - 설치 가이드 및 사용법 섹션 추가
```

### style (코드 스타일 혹은 포맷 등에 관한 커밋)

```
[#29] style: 코드 포맷 수정 - Prettier 적용하여 코드 포맷 수정
```

### refactor (코드 리팩토링에 대한 커밋)

```
[#12] refactor: 로그인 로직 리팩토링 - 로그인 로직을 별도 모듈로 분리하여 코드 가독성 향상
```

### test (테스트 코드 수정에 대한 커밋)

```
[#45] test: 로그인 기능 테스트 추가 - 로그인 성공/실패 케이스에 대한 테스트 추가
```

### perf (성능 개선에 대한 커밋)

```
[#67] perf: 이미지 로딩 성능 개선 - 비동기 이미지 로딩 및 캐싱 적용으로 스크롤 성능 향상
```

## Pull Request Template
```
# 제목
[PR 제목을 입력하세요]

# 설명
이 PR은 다음과 같은 변경 사항을 포함합니다:
- [변경 사항 1]
- [변경 사항 2]

# 관련 이슈
Closes #1

# 변경 사항
- [변경된 내용 설명]

# 테스트 계획
1. [테스트 단계 1]
2. [테스트 단계 2]

# 검토자에게 할 말
- [리뷰어에게 전할 내용]

```

**Comment by yeniful** [comment]
이슈에서 브랜치를 바로 만들 수도 있군요!
깃허브가 익숙하지 않은 분들께 너무 좋은 자료가 될 것 같아요!!
